File: dynabuffers-python/dynabuffers/ast/FieldType.py
from dynabuffers.api.ISerializable import ISerializable, ByteBuffer
from dynabuffers.ast.structural import FieldOptions


class FieldTypeOptions():
    def __init__(self, name: str, datatype: ISerializable, options: FieldOptions, annotations, defaultVal):
        self.name = name
        self.datatype = datatype
        self.options = options
        self.annotations = annotations
        self.defaultVal = defaultVal


class FieldType(ISerializable):

    def __init__(self, options: FieldTypeOptions):
        self.options = options

    def size(self, value, registry):
        return 1 + len(str(value).encode("utf-8"))

    def serialize(self, value, buffer: ByteBuffer, registry):
        if self.options.options:
            registry.addNotification("deprecated field " + self.options.name + " used")

        for annotation in list(map(lambda x: registry.resolveAnnotation(x.options.name, x.options.values), self.options.annotations)):
            annotation.validate(self.options.name, value)

        self.options.datatype.serialize(value, buffer, registry)

    def deserialize(self, buffer, registry):
        return self.options.datatype.deserialize(buffer, registry)

File: dynabuffers-java/src/main/kotlin/dynabuffers/ast/FieldType.kt
package dynabuffers.ast

import dynabuffers.api.IRegistry
import dynabuffers.api.ISerializable
import dynabuffers.api.IType
import dynabuffers.ast.structural.Annotation
import dynabuffers.ast.structural.FieldOptions
import java.nio.ByteBuffer

data class FieldType(val options: FieldTypeOptions) : IType, ISerializable {

    override fun size(value: Any, registry: IRegistry) = options.dataType.size(value, registry)

    override fun serialize(value: Any, buffer: ByteBuffer, registry: IRegistry) {
        if (options.options.isDeprecated())
            registry.addNotification("deprecated field ${options.name} used.")

        options.annotations.forEach { registry.resolveAnnotation(it).validate(options.name, value) }

        options.dataType.serialize(value, buffer, registry)
    }

    override fun deserialize(buffer: ByteBuffer, registry: IRegistry): Any {
        if (options.options.isDeprecated())
            registry.addNotification("deprecated field ${options.name} used.")

        return options.dataType.deserialize(buffer, registry)
    }

    data class FieldTypeOptions(val name: String, val annotations: List<Annotation>, val dataType: ISerializable, val options: FieldOptions, val defaultVal: Any?)

}
